source/codegen/templates/proto_helpers.mako

```diff
@@ -79,23 +85,38 @@ message ${common_helpers.snake_to_pascal(function)}Request {
 % for parameter in input_parameters:
 <%
   parameter_name = common_helpers.camel_to_snake(parameter["name"])
-  grpc_field_number = proto_helpers.generate_parameter_field_number(parameter, used_indexes)
+  is_array = common_helpers.is_array(parameter["type"])
   if 'grpc_type' in parameter:
     parameter_type = parameter['grpc_type']
   else:
-    parameter_type = proto_helpers.determine_function_parameter_type(parameter, service_class_prefix)
+    parameter_type = proto_helpers.get_grpc_type_from_ivi(parameter["type"], is_array, service_class_prefix)
 %>\
 %   if common_helpers.is_enum(parameter):
 <%
-  is_array = common_helpers.is_array(parameter["type"])
-  non_enum_type = proto_helpers.get_grpc_type_from_ivi(parameter["type"], is_array, service_class_prefix)
+  enum_parameter_type = None
+  mapped_type = None
+  if parameter.get('enum', None):
+    enum_parameter_type = f"repeated {parameter['enum']}" if is_array else parameter['enum']
+    grpc_enum_field_number = proto_helpers.generate_parameter_field_number(parameter, used_indexes)
+  if parameter.get('mapped-enum', None):
+    mapped_type = f"repeated {parameter['mapped-enum']}" if is_array else parameter['mapped-enum']
+    mapped_type = mapped_type if mapped_type.endswith('Mapped') else f'{mapped_type}Mapped'
+    grpc_mapped_field_number = proto_helpers.generate_parameter_field_number(parameter, used_indexes, "_mapped")
   grpc_raw_field_number = proto_helpers.generate_parameter_field_number(parameter, used_indexes, "_raw")
 %>\
   oneof ${parameter_name}_enum {
-    ${parameter_type} ${parameter_name} = ${grpc_field_number};
-    ${non_enum_type} ${parameter_name}_raw = ${grpc_raw_field_number};
+%   if enum_parameter_type:
```

Comment:
Is there any way this conditional can collapse to:

`${parameter_type} ${parameter_name} = ${field_number};`

And the conditional code above (or python helper code) figures out the correct values for each of those?